odd_collector_azure/adapters/azure_sql/repository.py

```diff
@@ -0,0 +1,176 @@
+        )
+        self._connection: pyodbc.Connection = pyodbc.connect(self.connection_string)
+        self._cursor: pyodbc.Cursor = self._connection.cursor()
+
+    def get_tables(self) -> Iterable[Table]:
+        """Get table metadata."""
+        query = TABLES_METADATA_QUERY.format(database=self.config.database)
+        result = self._fetch_all(query)
+
+        for row in result:
+            table = Table(
+                name=row['table_name'],
+                columns=self.get_columns(row.get('table_name')),
+                create_date=row['create_date'].replace(tzinfo=pytz.utc).isoformat(),
+                modify_date=row['modify_date'].replace(tzinfo=pytz.utc).isoformat(),
+                row_count=row['row_count'],
+                description=row['type_desc'],
+            )
+            yield table
+
+    def get_columns(self, table_name: str) -> List[Dict]:
+        """Get columns metadata."""
+        query = COLUMNS_METADATA_QUERY.format(table_name=table_name)
+        columns = self._fetch_all(query)
+        return columns
+
+    def get_views(self) -> Iterable[View]:
+        """Get views metadata."""
+        views_data = self._fetch_all(VIEW_METADATA_QUERY)
+        dependencies = self.get_view_dependencies()
+
+        views: Dict[str, View] = {
+            view.get('table_name'): View(
+                name=view.get('table_name'),
+                columns=self.get_columns(view.get('table_name')),
+                view_definition=view.get('VIEW_DEFINITION'),
+                description='',
+                upstream=[],
+                downstream=[],
+            )
+            for view in views_data
+        }
+
+        for dep in dependencies:
+            if dep.name in views:
+                views[dep.name].upstream.append(dep)
+
+                if (
+                    dep.referenced_type == DependencyType.VIEW
+                    and dep.referenced_name in views
+                ):
+                    views[dep.referenced_name].downstream.append(dep)
+        return iter(views.values())
+
+    def get_view_dependencies(self) -> Iterable[Dependency]:
+        """Get view dependencies."""
+        dependencies = self._fetch_all(VIEW_DEPENDENCIES_QUERY)
+        for row in dependencies:
+            yield Dependency(
+                name=row.get('name'),
+                referenced_owner='',
+                referenced_name=row.get('referenced_entity_name'),
+                referenced_type=DependencyType(row.get('referenced_type').strip())
+            )
+
+    def _fetch_all(self, query: str) -> List[dict]:
+        """Fetch data from DB and convert it to the dictionary as {column_name: value}."""
+        self._cursor.execute(query)
+        columns = [column[0] for column in self._cursor.description]
+        results = []
+        for row in self._cursor.fetchall():
+            results.append(dict(zip(columns, row)))
+        return results
+
+    def __del__(self):
+        self._connection.close()
```

Comment:
it isn't a good idea to have opened connection for whole object lifetime. We do not know when Python (sdk) will delete it.

Delete connection and cursor after each meaningful operation, in you case - load all data
```
create connection, cursor
   >>> get_tables()
   >>> get_columns()
   ... other operations
delete connection, cursor  

```

Reply:
Done